### PR TITLE
Mot3 5372 add is motion safety property

### PR DESCRIPTION
**Pull request by lucia-teruel**
### Description

This pull request introduces a unified way to determine if a safety function is motion-related by adding an `is_motion()` method to the `SafetyFunction` base class and overriding it in relevant subclasses. It also adds comprehensive tests to verify the correct classification of each safety function as motion or non-motion.

### Type of change

* Added a default `is_motion()` method and a `_custom_motion_rules()` hook to the `SafetyFunction` base class, allowing subclasses to specify if they are motion-related.
* Overrode `is_motion()` in all relevant safety function subclasses (`SS1Function`, `SS2Function`, `SOutFunction`, `SVFunction`, `SafeHomingFunction`, `SLSFunction`, `SSRFunction`, `SLPFunction`, `SDIFunction`, `SLIFunction`, etc.) to return `True` or implement custom logic as appropriate. [[1]](diffhunk://#diff-1e59b948fe520bf99bdd84f46cfddcb6e62c9f7a219845e20fa3fc9e231cb1a0R450-R471) [[2]](diffhunk://#diff-1e59b948fe520bf99bdd84f46cfddcb6e62c9f7a219845e20fa3fc9e231cb1a0R524-R532) [[3]](diffhunk://#diff-1e59b948fe520bf99bdd84f46cfddcb6e62c9f7a219845e20fa3fc9e231cb1a0R585-R593) [[4]](diffhunk://#diff-1e59b948fe520bf99bdd84f46cfddcb6e62c9f7a219845e20fa3fc9e231cb1a0R653-R661) [[5]](diffhunk://#diff-1e59b948fe520bf99bdd84f46cfddcb6e62c9f7a219845e20fa3fc9e231cb1a0R671-R679) [[6]](diffhunk://#diff-1e59b948fe520bf99bdd84f46cfddcb6e62c9f7a219845e20fa3fc9e231cb1a0R694-R702) [[7]](diffhunk://#diff-1e59b948fe520bf99bdd84f46cfddcb6e62c9f7a219845e20fa3fc9e231cb1a0R727-R735) [[8]](diffhunk://#diff-1e59b948fe520bf99bdd84f46cfddcb6e62c9f7a219845e20fa3fc9e231cb1a0R763-R771) [[9]](diffhunk://#diff-1e59b948fe520bf99bdd84f46cfddcb6e62c9f7a219845e20fa3fc9e231cb1a0R800-R808) [[10]](diffhunk://#diff-1e59b948fe520bf99bdd84f46cfddcb6e62c9f7a219845e20fa3fc9e231cb1a0R826-R834) [[11]](diffhunk://#diff-1e59b948fe520bf99bdd84f46cfddcb6e62c9f7a219845e20fa3fc9e231cb1a0R861-R869)

### Tests

* Added new tests for `is_motion()` to verify the motion-related status of all safety function types, including edge cases for functions like `SS1Function` where the result depends on parameters. [[1]](diffhunk://#diff-f301188661f93fbe0d4ac4859c929bfb53f20d719394bea6324cb0c827392df7R241-R251) [[2]](diffhunk://#diff-f301188661f93fbe0d4ac4859c929bfb53f20d719394bea6324cb0c827392df7R271-R281) [[3]](diffhunk://#diff-f301188661f93fbe0d4ac4859c929bfb53f20d719394bea6324cb0c827392df7R321-R331) [[4]](diffhunk://#diff-f301188661f93fbe0d4ac4859c929bfb53f20d719394bea6324cb0c827392df7R390-R405) [[5]](diffhunk://#diff-f301188661f93fbe0d4ac4859c929bfb53f20d719394bea6324cb0c827392df7R429-R439) [[6]](diffhunk://#diff-f301188661f93fbe0d4ac4859c929bfb53f20d719394bea6324cb0c827392df7R464-R485) [[7]](diffhunk://#diff-f301188661f93fbe0d4ac4859c929bfb53f20d719394bea6324cb0c827392df7R509-R585)

These changes improve the clarity and consistency of how motion-related safety functions are identified in the codebase, and ensure correct behavior through thorough testing.

### Documentation

Please update the documentation.

- [x] Update docstrings of every function, method or class that change.
- [x] Add the changes at the `[Unreleased]` section of the [CHANGELOG](CHANGELOG.md).
- [x] Comment tests.

### CODE REVIEW

File: ingeniamotion/fsoe_master/safety_functions.py [hunk 1]
```diff
@@ -429,6 +447,33 @@ def activated_by(self) -> list[SafetyFunction]:
         )
         return sf_list
 
+    @override
+    def is_motion(self) -> bool:
```
Comment (GullinBustin) [inline ingeniamotion/fsoe_master/safety_functions.py:456]: You don't need to write the docstring again and again

File: ingeniamotion/fsoe_master/safety_functions.py [hunk 2]
```diff
@@ -429,6 +447,33 @@ def activated_by(self) -> list[SafetyFunction]:
...
+            True if the safety function is a motion-related function, False otherwise.
+        """
+        # If Function is in Deceleration mode -> it is a motion function,
+        # otherwise it is a stop function without motion (SS1 with time to STO but no deceleration)
+        if self.deceleration_limit is None:
+            raise ValueError("SS1Function is missing deceleration_limit parameter.")
+        return bool(self.deceleration_limit.get())
+
+    @override
+    def custom_motion_rules(self) -> bool:
```
Comment (GullinBustin) [inline ingeniamotion/fsoe_master/safety_functions.py]: I don't understand what this function is for.

Comment (lucia-teruel) [reply ingeniamotion/fsoe_master/safety_functions.py]: There might be some SafetyFunctions (like SS1) who's `is_motion` is not always the same, but it depends on Parameters or other facts (such as a register for the SS1 case).
Thought it was a good idea to make this distinction, as these should not be blocked on the ml3 Safety Functions widget page.

File: ingeniamotion/fsoe_master/safety_functions.py [hunk 3]
```diff
@@ -429,6 +452,20 @@ def activated_by(self) -> list[SafetyFunction]:
         )
         return sf_list
 
+    @override
+    def is_motion(self) -> bool:
+        # If Function is in Deceleration mode -> it is a motion function,
+        # otherwise it is a stop function without motion (SS1 with time to STO but no deceleration)
+        if self.deceleration_limit is None:
+            raise ValueError("SS1Function is missing deceleration_limit parameter.")
```
Comment (polfeliu) [inline ingeniamotion/fsoe_master/safety_functions.py]: i don't agree with this handling.

deceleration limit is optional because it does not agree with phase 1. On phase 1 there was no motion, thus, this parameter did not exist, so this should simply return False

File: ingeniamotion/fsoe_master/safety_functions.py (hunk 2 above)
Comment (julieta-prieto) [reply ingeniamotion/fsoe_master/safety_functions.py]: if it depend on other things, then why not return a different thing in `is_motion` rather than creating another method for it?

**Review by polfeliu — APPROVED** [review]

File: ingeniamotion/fsoe_master/safety_functions.py (hunk 2 above)
Comment (lucia-teruel) [reply ingeniamotion/fsoe_master/safety_functions.py]: I tried to follow a similar structure to the "is_active" property...

I also need to differentiate all of these cases:
- NO Motion SafetyFunction
- YES Motion SafetyFunction
- DEPENDING Motion SafetyFunction (Its "is_motion" status depends on Parameters/Other rules)

Comment (polfeliu) [reply ingeniamotion/fsoe_master/safety_functions.py]: Theres a possible other pattern if you are interested. Maybe it's overkill and I have note tested

```python
class SafetyFunctionCondition:
	
    def __init__(self, func):
        self.func = func

	def on(self, sf: SafetyFunction):
		return self.func(sf)

class SafetyFunction:

	@classmethod
	def is_motion(cls) -> Union[bool, SafetyFunctionCondition]:
		return False
	

class CustomSafetyFunction:

	def configured_as_motion(self) -> bool:
		self.parameter == 42
		return True

	@classmethod
	def is_motion(cls) -> Union[bool, SafetyFunctionCondition]:
		return SafetyFunctionCondition(cls.configured_as_motion)


my_safety_function = CustomSafetyFunction()

is_motion = my_safety_function.is_motion()

if isinstance(is_motion, bool):
    if is_motion:
        print("This function is always a motion function.")
    else:
		print("This function is never a motion function.")
elif isinstance(is_motion, SafetyFunctionCondition):
    if is_motion.on(my_safety_function):
        print("This function is a motion function based on the configuration.")
    else:
        print("This function is not a motion function based on the configuration.")

	
```

**Review by GullinBustin — APPROVED** [review]
